src/dtx/dtx_common.c

```diff
@@ -0,0 +1,632 @@
+
+	return result > 0 ? 0 : result;
+}
+
+int
+dtx_conflict(daos_handle_t coh, struct dtx_handle *dth, uuid_t po_uuid,
+	     uuid_t co_uuid, struct dtx_conflict_entry *dces, int count,
+	     uint32_t version)
+{
+	daos_unit_oid_t		*oid = &dth->dth_oid;
+	struct dtx_id		*commit_ids = NULL;
+	struct dtx_entry	*abort_dtes = NULL;
+	int			 commit_cnt = 0;
+	int			 abort_cnt = 0;
+	int			 rc = 0;
+	int			 i;
+
+	D_ASSERT(dth->dth_leader);
+
+	D_ALLOC_ARRAY(commit_ids, count);
+	if (commit_ids == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	D_ALLOC_ARRAY(abort_dtes, count);
+	if (abort_dtes == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	for (i = 0; i < count; i++) {
+		int	j;
+		bool	skip = false;
+
+		for (j = 0; j < i; j++) {
+			if (daos_dti_equal(&dces[i].dce_xid,
+					   &dces[j].dce_xid)) {
+				skip = true;
+				break;
+			}
+		}
+
+		if (!skip) {
+			rc = vos_dtx_lookup_cos(coh, oid, &dces[i].dce_xid,
+						dces[i].dce_dkey, true);
+			if (rc != -DER_NONEXIST)
+				goto found;
+
+			rc = vos_dtx_lookup_cos(coh, oid,
+						&dces[i].dce_xid,
+						dces[i].dce_dkey, false);
+			if (rc != -DER_NONEXIST)
+				goto found;
+
+			rc = vos_dtx_check_committable(coh, NULL,
+						&dces[i].dce_xid,
+						dces[i].dce_dkey, true);
+			if (rc == DTX_ST_COMMITTED)
+				rc = 0;
+			else if (rc >= 0)
+				rc = -DER_NONEXIST;
+
+found:
+			if (rc == 0) {
+				daos_dti_copy(&commit_ids[commit_cnt++],
+					      &dces[i].dce_xid);
+				continue;
+			}
+
+			if (rc == -DER_NONEXIST) {
+				daos_dti_copy(&abort_dtes[abort_cnt].dte_xid,
+					      &dces[i].dce_xid);
+				abort_dtes[abort_cnt++].dte_oid = *oid;
+				continue;
+			}
+
+			goto out;
+		}
+	}
+
+	if (commit_cnt > 0) {
+		struct dtx_id	*dti_cos;
+		int		 dti_cos_count;
+
+		/* Append the committable DTXs' ID to the CoS list. */
+		dti_cos_count = dth->dth_dti_cos_count + commit_cnt;
+		D_ALLOC_ARRAY(dti_cos, dti_cos_count);
+		if (dti_cos == NULL)
+			D_GOTO(out, rc = -DER_NOMEM);
+
+		if (dth->dth_dti_cos != NULL) {
+			memcpy(dti_cos, dth->dth_dti_cos,
+			       sizeof(struct dtx_id) * dth->dth_dti_cos_count);
+
+			D_FREE(dth->dth_dti_cos);
+		}
+
+		memcpy(dti_cos + dth->dth_dti_cos_count, commit_ids,
+		       sizeof(struct dtx_id) * commit_cnt);
+		dth->dth_dti_cos_count = dti_cos_count;
+		dth->dth_dti_cos = dti_cos;
+	}
+
+	if (abort_cnt > 0) {
+		rc = dtx_abort(po_uuid, co_uuid, abort_dtes, abort_cnt,
+				version);
+		if (rc == -DER_NONEXIST)
+			rc = 0;
+	}
+
+out:
+	if (commit_ids != NULL)
+		D_FREE(commit_ids);
+	if (abort_dtes != NULL)
+		D_FREE(abort_dtes);
+
+	return rc > 0 ? 0 : rc;
+}
+
+int
+dtx_batched_commit_register(struct ds_cont_hdl *hdl)
+{
+	struct ds_cont			*cont = hdl->sch_cont;
+	struct dtx_batched_commit_args	*dbca;
+	d_list_t			*head;
+
+	D_ASSERT(cont != NULL);
+
+	head = &dss_get_module_info()->dmi_dtx_batched_list;
+	d_list_for_each_entry(dbca, head, dbca_link) {
+		if (uuid_compare(dbca->dbca_cont_hdl->sch_cont->sc_uuid,
+				 cont->sc_uuid) == 0)
```

Comment:
could it be being closed? What will happen if it is

Reply:
what's mean? the dtx_batched_commit_register only will be called when open the container. When close, the sponsor will call dtx_batched_commit_deregister().